src/main/kotlin/org/koitharu/kotatsu/parsers/site/en/LineWebtoonsParser.kt

```diff
@@ -0,0 +1,426 @@
+						description = null,
+						state = null,
+						source = source,
+					)
+				}
+		} else {
+			val result = makeRequest("/lineWebtoon/webtoon/challengeGenreTitleList.json?genre=${genre}&sortOrder=${sortOrderStr}&startIndex=${offset+1}&pageSize=20")
+
+			val genres = result.getJSONObject("genreList")
+				.getJSONArray("challengeGenres")
+				.mapJSON { jo -> parseTag(jo) }
+				.associateBy { tag -> tag.key }
+
+			result
+				.getJSONObject("titleList")
+				.getJSONArray("titles")
+				.mapJSON { jo ->
+					val titleNo = jo.getLong("titleNo")
+
+					Manga(
+						id = generateUid(titleNo),
+						title = jo.getString("title"),
+						altTitle = null,
+						url = "$titleNo",
+						publicUrl = "https://${domain}/en/canvas/a/list?title_no=${titleNo}",
+						rating = jo.getDouble("starScoreAverage").toFloat() / 10f,
+						isNsfw = jo.getBoolean("ageGradeNotice"),
+						coverUrl = "https://$staticDomain${jo.getString("thumbnail")}",
+						largeCoverUrl = if (jo.has("thumbnailVertical")) {
+							"https://$staticDomain${jo.getString("thumbnailVertical")}"
+						} else { null },
+						tags = setOf(genres[jo.getString("representGenre")]!!),
+						author = jo.getString("writingAuthorName"),
+						description = jo.getString("synopsis"),
+						// I don't think the API provides this info
+						state = null,
+						source = source,
+					)
+				}
+		}
+
+		return manga
+	}
+
+	override suspend fun getPages(chapter: MangaChapter): List<MangaPage> {
+		val (titleNo, episodeNo) = chapter.url.splitTwoParts('-')!!
+
+		return makeRequest("/lineWebtoon/webtoon/challengeEpisodeInfo.json?v=2&titleNo=${titleNo}&episodeNo=${episodeNo}")
+			.getJSONObject("episodeInfo")
+			.getJSONArray("imageInfo")
+			.mapJSONIndexed() { i, jo ->
+				MangaPage(
+					id = generateUid("$titleNo-$episodeNo-$i"),
+					url = "https://$staticDomain${jo.getString("url")}",
+					preview = null,
+					source = source,
+				)
+			}
+	}
+
+	private fun parseTag(jo: JSONObject): MangaTag {
+		return MangaTag(
+			title = jo.getString("name"),
+			key = jo.getString("code"),
+			source = source,
+		)
+	}
+
+	override suspend fun getTags(): Set<MangaTag> {
+		return makeRequest("/lineWebtoon/webtoon/challengeGenreList.json")
+			.getJSONObject("genreList")
+			.getJSONArray("challengeGenres")
+			.mapJSONToSet { jo -> parseTag(jo) }
+	}
+
+	private suspend fun makeRequest(url: String): JSONObject {
+		val resp = webClient.httpGet(finalizeUrl(url))
+		val message = resp.parseJson().getJSONObject("message")
+		if (resp.isSuccessful) {
+			return message.getJSONObject("result")
+		} else {
+			// TODO: handle 404 and some other error codes
+			val code = message.getInt("code")
+			val errorMessage = message.getString("message")
+			throw ParseException("Api error (code=$code): $errorMessage", url)
+		}
+	}
+
+	private fun finalizeUrl(url: String): String {
+		val urlWithHost = "https://${apiDomain}$url"
+		val uri = URI(urlWithHost)
```

Comment:
It might be better to use `HttpUrl` instead of `URI`

Reply:
Oh, I was wondering why this class didn't parse the query string :sweat_smile: